contract/migrations/12.0.4.0.0/post-migration.py

```diff
@@ -0,0 +1,210 @@
+    account_analytic_invoice_line"""
+    return (
+        'contract_id'
+        if openupgrade.column_exists(
+            cr, 'account_analytic_invoice_line', 'contract_id'
+        )
+        else 'analytic_account_id'
+    )
+
+
+def _copy_contract_table(cr):
+    contract_fields = []
+    for field in CONTRACT_FIELDS:
+        if openupgrade.column_exists(cr, 'account_analytic_account', field):
+            contract_fields.append(field)
+    contract_field_name = _get_contract_field_name(cr)
+    openupgrade.logged_query(
+        cr,
+        "INSERT INTO contract_contract ("
+        + ', '.join(contract_fields)
+        + ") "
+        + "SELECT "
+        + ', '.join(contract_fields)
+        + " FROM account_analytic_account "
+        + "WHERE id in ( SELECT DISTINCT "
+        + contract_field_name
+        + " FROM "
+        + "account_analytic_invoice_line)",
+    )
+    if openupgrade.column_exists(cr, 'account_invoice', 'old_contract_id_tmp'):
+        openupgrade.logged_query(
+            cr,
+            """
+            UPDATE account_invoice
+            SET old_contract_id = old_contract_id_tmp
+            """,
+        )
+
+    # Move contract attachments
+    openupgrade.logged_query(
+        cr,
+        "UPDATE ir_attachment SET res_model='contract.contract'"
+        + "WHERE res_model='account.analytic.account' "
+        + "AND res_id IN ( SELECT DISTINCT "
+        + contract_field_name
+        + " FROM account_analytic_invoice_line)",
+    )
+    # Move contract messages
+    openupgrade.logged_query(
+        cr,
+        "UPDATE mail_message SET model='contract.contract'"
+        + "WHERE model='account.analytic.account' "
+        + "AND res_id IN ( SELECT DISTINCT "
+        + contract_field_name
+        + " FROM account_analytic_invoice_line)",
+    )
+    # Move contract followers
+    openupgrade.logged_query(
+        cr,
+        "UPDATE mail_followers SET res_model='contract.contract'"
+        + "WHERE res_model='account.analytic.account' "
+        + "AND res_id IN ( SELECT DISTINCT "
+        + contract_field_name
+        + " FROM account_analytic_invoice_line)",
+    )
+
+
+def _copy_contract_line_table(cr):
+    contract_line_fields = []
+    contract_field_name = _get_contract_field_name(cr)
+    for field in CONTRACT_LINE_FIELDS:
+        if openupgrade.column_exists(
+            cr, 'account_analytic_invoice_line', field
+        ):
+            contract_line_fields.append(field)
+    account_analytic_invoice_line_fields = contract_line_fields.copy()
+    contract_line_fields.append('contract_id')
+    account_analytic_invoice_line_fields.append(contract_field_name)
+    contract_line_fields.append('analytic_account_id')
+    account_analytic_invoice_line_fields.append(contract_field_name)
+    contract_line_fields.append('active')
+    account_analytic_invoice_line_fields.append('true')
+
+    openupgrade.logged_query(
+        cr,
+        "INSERT INTO contract_line ("
```

Comment:
@sbejaoui  what is the reason for inserting records instead of renaming the table? I'm just experimenting now with the renaming to see problems, but I don't see in theory anyone.

Reply:
As we are not renaming the `account_analytic_account` table. Renaming the `account_analytic_invoice_line` table in pre-migration trigger foreign-key constraint to `contract_contract` table as it's not created yet.

With @ThomasBinsfeld , we think we had two solution to resolve this.
1- Copy the create the `contract_contract` and copy in it the `account_analytic_account` table in the pre-migration script. Rename `account_analytic_invoice_line` table to contract_line.
2- Let odoo create the `contract_contract` and the `contract_line` tables and use insert to fill the data.

Do you think we can resolve the foreign-key constraint issue differently?

Reply:
Yes, I think so. Usually the best option is to directly remove that FK constraint in pre and let Odoo to do rebuild it, but I'm trying right now. What are the benefits of not needed to move data?:

1. Speed: if you have a lot of records, insert operation takes time. As this is by SQL, it's not too much, but it's something.
2. Additional modules that added in v11 fields to this table, will need to create extra migration scripts for moving these fields values to the new table with this approach. If we simply rename the table, SQL columns will be there and then no migration script will be needed.
3. Avoid FK constraints.

What I'm going to try is to create table copying columns as is. Having extra columns there (coming from account_analytic_account table) is not problem, as they can be cleaned afterwards.